crates/sui/src/upgrade_compatibility.rs

```diff
@@ -1509,6 +1545,65 @@ fn enum_ability_mismatch_diag(
     Ok(diags)
 }
 
+fn enum_variant_field_error(
+    old_variant: &Variant,
+    new_variant: &Variant,
+    variant_loc: Loc,
+    def_loc: Loc,
+) -> (DiagnosticInfo, Vec<String>) {
+    if old_variant.fields.len() != new_variant.fields.len() {
+        return (
+            Declarations::FieldMismatch.into(),
+            vec![format!(
+                "Mismatched variant field count, expected {}, found {}.",
+                old_variant.fields.len(),
+                new_variant.fields.len()
+            )],
+        );
+    }
+
+    match (
+        old_variant.name != new_variant.name,
+        old_variant.fields != new_variant.fields,
+    ) {
+        (true, true) => (
+            Enums::VariantMismatch.into(),
```

Comment:
Curious why `into()` is required here. Is there a From for Enums to Diagnostic?
Maybe the better question is why is this not part of Diagnostic?

Reply:
all the branches need to be the same type and since the last branch uses a different category it ends up having a different type without `into()`

the codes! implementation creates a category enum and adds individual errors under that enum each category is its own type that can into() a diagnostic info.